src/libs/commonlib.cpp

```diff
@@ -839,9 +983,115 @@ Namespace* CoreIRLoadLibrary_commonlib(Context* c) {
     lb_args
   );
   lb->addDefaultGenArgs({{"has_valid",Const::make(c,false)}});
-  lb->addDefaultGenArgs({{"is_last_lb",Const::make(c,true)}}); // asserted false to not create register chain
 
   lb->setGeneratorDefFromFun([](Context* c, Values genargs, ModuleDef* def) {
+      bool has_valid = genargs.at("has_valid")->get<bool>();
+      bool is_last_lb= true;
+      Type* in_type  = genargs.at("input_type")->get<Type*>();
+      Type* out_type = genargs.at("output_type")->get<Type*>();
+      Type* img_type = genargs.at("image_type")->get<Type*>();
+
+      // create a linebuffer
+      Values lb_args = {
+            {"input_type", Const::make(c, in_type)},
+            {"image_type", Const::make(c, img_type)},
+            {"output_type", Const::make(c, out_type)},
+            {"has_valid", Const::make(c, has_valid)},
+            {"is_last_lb", Const::make(c, is_last_lb)}
+      };
+
+      def->addInstance("lb_recurse",
+                       "commonlib.linebuffer_recursive",
+                       lb_args);
+
+      // connect linebuffer to this generator's ports
+      def->connect("self.in", "lb_recurse.in");
+      def->connect("self.reset", "lb_recurse.reset");
+      def->connect("self.wen", "lb_recurse.wen");
+      if (has_valid) {
+        def->connect("self.valid", "lb_recurse.valid");
+      }
+
+      // flip all of the outputs to the correct output port
+      vector<uint> in_dims = get_dims(in_type);
+      vector<uint> out_dims = get_dims(out_type);
+      vector<uint> img_dims = get_dims(img_type);
+      in_dims.erase(in_dims.begin()); 
+      out_dims.erase(out_dims.begin());
+      img_dims.erase(img_dims.begin());
+      uint num_dims = in_dims.size();
+
+      std::vector<std::pair<string,string>> out_pairs;
+      out_pairs.push_back({"lb_recurse.out", "self.out"});
+      for (int dim=num_dims-1; dim>=0; --dim) {
+        uint  inx =  in_dims[dim];
+        uint outx = out_dims[dim];
+
+        std::vector<std::pair<string,string>> out_temp;
+        out_temp.reserve(out_pairs.size() * outx);
+        for (uint i=0; i<outx; ++i) {
+          for (auto out_pair : out_pairs) {
+            string source = out_pair.first;
+            string sink = out_pair.second;
+            uint iflip = inverted_index(outx, inx, i);
+            out_temp.push_back({source + "." + to_string(i),
+                    sink + "." + to_string(iflip)});
+            
+          }
+        }
+        out_pairs = out_temp;
+      }
```

Comment:
This loop is also another good candidate for a helper function (e.g. `make_out_pairs`), also a comment explaining what this is doing would be good.